### ISSUE TITLE
Blank lines

### ISSUE THREAD

**Issue by jklmnn** [Componolit/ada-style#3]
As with the discussion at https://github.com/Componolit/RecordFlux/pull/198 we should also think about a policy about where to use blank lines (except for the rule that there shouldn't be multiple blank lines in a row).

**Comment by senier** [comment]
Ada Style Guide:

# 2.2.6 Blank Lines

## guideline
- Use blank lines to group logically related lines of text (NASA 1987).

## rationale
When blank lines are used in a thoughtful and consistent manner, sections of related code
are more visible to readers.

## automation notes

Automatic formatters do not enforce this guideline well because the decision on where to insert blank lines is a semantic one. However, many formatters have the ability to leave existing blank lines intact. Thus, you can manually insert the lines and not lose the effect when you run such a formatter.

**Comment by jklmnn** [comment]
My personal preferences on this topic:

## Packages

- One blank line after `is` and before `end` except for pragmas at the beginning of the package
 - Scalar types and array types can be grouped, records should be separated by a blank line
- Single line function and procedure declarations can be grouped, especially multiple declarations of generic entities.
 - Function and procedure declarations that use more than 2 lines (including contracts) should be separated by a blank line.
- Procedure bodies should always be separated (except multiple null procedures)
- Expression functions might be grouped if they are single line including their expression
```Ada
package P
is
   pragma Pure;

   type I is new Integer;
   type U is new Natural;

   type R is record
      I : Integer;
   end record;

   procedure P1;
   procedure P2;

   procedure P3 (A : Integer;
                 B : Integer);

   function F (I : Integer) return Integer with
      Pre => I > 0;

end P;

package body P
is

   procedure P1 is null;
   procedure P2 is null;

   procedure P3 (A : Integer;
                 B : Integer)
   is
   begin
      null;
   end P3;

   function F (I : Integer) return Integer is (-I);

end P;
```

## Subprograms

- Blank lines should only be used to separate larger blocks of code inside a subprogram
- Blank lines might be used to separate local subprograms in the declarative part
- Blank lines should never be after `is`, before and after `begin` and before `end`

```Ada
procedure P
is
   I : Integer;
begin
   I := 1;
end P;

procedure P
is
   procedure Q
   is
   begin
      null;
   end Q;
begin
   null;
end P;

procedure P
is
   procedure Q;
   procedure R;

   procedure Q
   is
   begin
      null;
   end Q;

   procedure R
   is
   begin
      null;
   end R;
begin
   null;
end P;

procedure P
is
begin
   loop
      null;
   end loop;
end P;

procedure P
is
begin
   loop;
      null;
   end loop;

   loop;
      null;
   end loop;
end P;
```

**Comment by senier** [comment]
I like the suggestion very much so far.

* Regarding packages:

   We should include examples for `with` statements at the beginning of packages. My preference is: no blank line at start of file, one black line between `with` and `package`.

   Do we want to allow blank lines between `with` statements, e.g. to separate external vs. project local imports or similar? In any case there should be a clear rule.

* Regarding subprograms:

   We should tighten the rules for when a blank line is permissible. E.g. no blank lines, except to separate multi-line local subprograms and loops. For variable declarations the same rules as for packages apply.

When comparing the package proposal and the subprogram proposal I think we should unify those as much as possible. This makes me wonder if we should either remove the blank after `is` / before `end` for packages or force a black line after `is` and before `end` for subprograms. I'm leaning towards removing it for brevity.

**Comment by jklmnn** [comment]
My rationale for having blank lines in packages is that most entities residing in packages have blank lines surrounding them while those in subprograms haven't. Subprograms and record declarations usually have blank lines around them to stand out as entities. `begin`, `is` and so on as well as code statements are too small by themselves to provide a reason to add a blank line. In my opinion this also clearly defines how the elaboration code has to be formatted:
```Ada
package body P
is

   procedure P1
   is
   begin
      null;
   end P1;

begin
   null;
end P;
```

Without blank lines at the begin and end of a package the code would be far less readable (note that this is a small example, it would have more impact if `P1` was a complex procedure):
```Ada
package body P
is
   procedure P1
   is
   begin
      null;
   end P1;
begin
   null;
end P;
```

**Comment by treiher** [comment]
Just a remark: I would suggest to keep the rules as simple as possible. Exceptions will make it hard to follow the rules. Especially the generation of code according to the style guide or the use of automated code formatting will be made more difficult. (I think we should really look into automated code formatting for Ada.)

Do we really need such exceptions:
- One blank line after is and before end __except for pragmas at the beginning of the package__
- Function and procedure declarations __that use more than 2 lines (including contracts)__ should be separated by a blank line.
 - Procedure bodies should always be separated __(except multiple null procedures)__

**Comment by senier** [comment]
> Just a remark: I would suggest to keep the rules as simple as possible. Exceptions will make it hard to follow the rules.

That was my point when saying "I think we should unify those as much as possible". IMHO we should sacrifice some potential style improvements for simplify of the rules. Otherwise it will be hard to maintain the style without tool support.

> I think we should really look into automated code formatting for Ada.

Absolutely! Ultimately I want `black` for Ada. This should be doable with libadalang, but it will still be significant effort.

**Comment by senier** [comment]
@jklmnn Can you try to reduce the rules to, say 3-4 and provide some examples so we can see what we end up with?

**Comment by jklmnn** [comment]
In this case I'd propose the following:

 - all entities in the declaration block of a package are separated by a blank line, including a line at the beginning and the end of the block
 - blank lines in the declarative part of subprograms, declare blocks etc. are only allowed before, between and after local subprogram bodies
 - in a begin-end block blank lines are allowed neither in packages nor subprograms

```Ada
package P with
   Pure
is

   pragma Assert (True);

   type T is new Integer;

   function Negative (V : T) return Boolean;

   function Positive (V : T) return Boolean;

   procedure P1;

end P;

package body P
is

   function Negative (V : T) return Boolean is (V < 0);

   function Positive (V : T) return Boolean
   is
   begin
      return V > 0;
   end Positive;

   procedure P1
   is
      procedure Local1;
      procedure Local2;

      procedure Local1
      is
      begin
         null;
      end Local1;

      procedure Local2
      is
      begin
         null;
      end Local2;
   begin
      null;
   end P1;

begin
   null;
end P;
```

About blank lines in begin-end blocks, I would stay with the rationale:

> Automatic formatters do not enforce this guideline well because the decision on where to insert blank lines is a semantic one. However, many formatters have the ability to leave existing blank lines intact. Thus, you can manually insert the lines and not lose the effect when you run such a formatter.

While I support avoiding blank lines there they sometimes can improve readability and since they carry semantic information they should be kept intact. I'm aware that this will always be subject of manual review then.

EDIT (@senier): Remove the pragma from the example as per #6
EDIT (@jklmnn): `pragma Assert` can be used in packages but is not a valid aspect identifier, added this as example

**Comment by senier** [comment]
I like the proposal. I just wonder in what situation a blank line between `begin`/`end` could encounter. As per rule 3 - never. We should check with our existing code base how readable the resulting code would be.

**Comment by jklmnn** [comment]
> I just wonder in what situation a blank line between begin/end could encounter. As per rule 3 - never. We should check with our existing code base how readable the resulting code would be.

That's the only exception I'd allow as described in my previous comment. Of course there's always the argument that if you have a subprogram that complex it could be multiple subprograms.

@senier I'd like to keep the pragma in the example to have a reference in case we need one. I've just chosen `Pure` as example. Maybe we should use on in the example that cannot be expressed as aspect. If no such pragma exists then we can omit it.

**Comment by senier** [comment]
> That's the only exception I'd allow as described in my previous comment. Of course there's always the argument that if you have a subprogram that complex it could be multiple subprograms.

Your comment above (and the quote from the style guide) only state that blank should be kept intact as they carry semantic information. I see no rule when to add a blank line. If we don't have that then blank lines can be added arbitrarily in begin/end blocks and we'll have a discussion each time.

**Comment by senier** [comment]
> @senier I'd like to keep the pragma in the example to have a reference in case we need one. I've just chosen `Pure` as example. Maybe we should use on in the example that cannot be expressed as aspect. If no such pragma exists then we can omit it.

I see. If you find such a pragma, please add it again.

**Comment by jklmnn** [comment]
> Your comment above (and the quote from the style guide) only state that blank should be kept intact as they carry semantic information. I see no rule when to add a blank line. If we don't have that then blank lines can be added arbitrarily in begin/end blocks and we'll have a discussion each time.

I know that there is no specific rule. I think this is a similar problem as indentation levels which we initially restricted quite strictly which then led to worse code. We still strive to keep the level low but having a too small hard limit might be counter productive. I think its the same with blank lines in these blocks. We should aim to never use them but if they improve the readability of a complex subprogram they should be allowed. I don't think a hard limit would be useful, what would it be, 1, 2, 3 every 50 lines of code?

**Comment by senier** [comment]
>  think this is a similar problem as indentation levels which we initially restricted quite strictly which then led to worse code. We still strive to keep the level low but having a too small hard limit might be counter productive. I think its the same with blank lines in these blocks. We should aim to never use them but if they improve the readability of a complex subprogram they should be allowed. I don't think a hard limit would be useful, what would it be, 1, 2, 3 every 50 lines of code?

I was not so much thinking of a quantitative restriction, but rather rules for when a blank line is permissible in a subprogram body. Looking through my own code, I seem to introduce blank lines before and after loops and if-blocks. I also tend to separate assignments/initialization from the following statements.

My point is just: If we cannot come up with actionable rules for subprogram bodies than we should not mandate anything. This seems what the Python and Ada style guides have done. Maybe its just too hard.

**Comment by jklmnn** [comment]
I think there is no easily applicable set of rules for this. Saying before or after if blocks and loops still leads to the question for the requirements such a block must fulfill to have a blank line. Also how would a automatic checker detect initialization sections? Also if we had only small sections but multiple of them it could increase the length of a subprogram that it won't fit on the screen anymore which IMO is much worse than having no blank lines.

I wouldn't mandate anything more specific than that blank lines should always be omitted and only used if they improve the code. The latter will always be subject to manual review.

**Comment by senier** [comment]
OK, let's leave it at that. How do we want to integrate the rules we gathered here into the style guide? IMHO it would be nice to augment the existing `Packages` and `Subprograms` instead of a separate section on blank lines.

**Comment by jklmnn** [comment]
Closed by #9.
